pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -394,199 +442,192 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
+					BoundObjectRef: &authenticationv1.BoundObjectReference{
+						Kind:       dockercfgSecretRef.Kind,
+						APIVersion: dockercfgSecretRef.APIVersion,
+						Name:       dockercfgSecretRef.Name,
+						UID:        dockercfgSecretRef.UID,
+					},
+				},
 			},
-		},
-		Type: v1.SecretTypeDockercfg,
-		Data: map[string][]byte{},
+			metav1.CreateOptions{},
+		)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to retrieve a token for SA '%s/%s': %v", saNamespace, saName, err)
 	}
-	klog.V(4).Infof("Creating dockercfg secret %q for service account %s/%s", dockercfgSecret.Name, serviceAccount.Namespace, serviceAccount.Name)
-
-	// prevent updating the DockerURL until we've created the secret
-	e.dockerURLLock.Lock()
-	defer e.dockerURLLock.Unlock()
 
-	dockercfg := credentialprovider.DockerConfig{}
-	for _, dockerURL := range e.dockerURLs {
-		dockercfg[dockerURL] = credentialprovider.DockerConfigEntry{
-			Username: "serviceaccount",
-			Password: string(tokenSecret.Data[v1.ServiceAccountTokenKey]),
-			Email:    "serviceaccount@example.org",
-		}
+	respToken := tokenResp.Status.Token
+	if len(respToken) == 0 {
+		return "", nil, fmt.Errorf("retrieved an empty token for SA '%s/%s'", saNamespace, saName)
 	}
-	dockercfgContent, err := json.Marshal(&dockercfg)
-	if err != nil {
-		return nil, false, err
-	}
-	dockercfgSecret.Data[v1.DockerConfigKey] = dockercfgContent
-	blockDeletion := false
-	ownerRef := metav1.NewControllerRef(tokenSecret, v1.SchemeGroupVersion.WithKind("Secret"))
-	ownerRef.BlockOwnerDeletion = &blockDeletion
-	dockercfgSecret.SetOwnerReferences([]metav1.OwnerReference{*ownerRef})
-
-	// Save the secret
-	createdSecret, err := e.client.CoreV1().Secrets(tokenSecret.Namespace).Create(context.TODO(), dockercfgSecret, metav1.CreateOptions{})
-	// If we cannot create this secret because the namespace it is being terminated isn't a thing we should fail and requeue a retry.
-	// Instead, we know that when a new namespace gets created, the serviceaccount will be recreated and we'll get a second shot at
-	// processing the serviceaccount.
-	if kapierrors.HasStatusCause(err, v1.NamespaceTerminatingCause) {
-		return nil, false, nil
-	}
-	return createdSecret, err == nil, err
+
+	return respToken, &tokenResp.Status.ExpirationTimestamp, nil
 }
 
-func (e *DockercfgController) syncDockercfgOwnerRefs(serviceAccount *v1.ServiceAccount) error {
-	for _, secretRef := range serviceAccount.Secrets {
-		secret, exists, err := e.secretCache.GetByKey(fmt.Sprintf("%s/%s", secretRef.Namespace, secretRef.Name))
-		if err != nil {
-			return err
+// createDockerPullSecret creates a dockercfg secret based on the token secret
+func (e *DockercfgController) createDockerPullSecret(ctx context.Context, serviceAccount *v1.ServiceAccount) (*v1.Secret, bool, error) {
+	pendingTokenName := serviceAccount.Annotations[PendingTokenAnnotation]
+
+	// If this service account has no record of a pending token name, record one
+	if len(pendingTokenName) == 0 {
```

Comment:
If you can build a valid dockerconfig secret synchronously, you'd be able to first build the secret and second directly update the serviceaccount with the pull secret and never need the intermediate annotation, right?